File: assets/scripts/cocos/M02StarWebSession.ts
// M02《点亮你温暖我》会话层 —— 纯逻辑(无 cc import), 由 Vitest 钉死。
// 管三件 domain 之上的东西: 电量(charges)、三板推进、每颗星的呈现态 + 胜负。
// cc 胶水层(M02StarWebView)只读 view()、把 tap 转给 tapNode(), 不自己算规则。

import { StarNetworkModel, type StarNetworkRules } from "../core/StarNetworkModel.ts";
import { boardGraph, type StarBoard, type StarWebConfig } from "../core/StarWebConfig.ts";

/** 单颗星的呈现态: 暗 / 衰减中(亮但支撑不足) / 冻结(亮且亮邻居达标) */
export type StarNodeStatus = "dark" | "decaying" | "frozen";
/** 一板的状态: 进行中 / 已全锁胜利 / 电量耗尽未胜 */
export type BoardStatus = "playing" | "won" | "exhausted";

export interface StarNodeView {
  id: string;
  x: number;
  y: number;
  life: number;
  lit: boolean;
  status: StarNodeStatus;
}

export interface StarWebView {
  boardId: string;
  boardIndex: number;
  boardCount: number;
  nodes: StarNodeView[];
  edges: [string, string][];
  chargesTotal: number;
  chargesLeft: number;
  status: BoardStatus;
}

export interface TapResult {
  accepted: boolean;
  reason?: "not_playing" | "unknown_node";
}

export class StarWebSession {
  private readonly boards: StarBoard[];
  private readonly rules: StarNetworkRules;
  private boardIndex = 0;
  private model: StarNetworkModel;
  private chargesUsed = 0;
  private status: BoardStatus = "playing";
  private readonly wonBoardIds = new Set<string>();

  constructor(config: StarWebConfig) {
    if (config.boards.length === 0) {
      throw new Error("StarWebSession requires at least one board");
    }
    this.boards = config.boards;
    this.rules = config.mechanic;
    this.model = this.buildModel(0);
  }

  /** 点一颗星 = 花一点电量走一拍。未知星不消耗电量、不推进；非进行中拒绝。 */
  tapNode(id: string): TapResult {
    if (this.status !== "playing") return { accepted: false, reason: "not_playing" };
    if (!this.model.step(id)) return { accepted: false, reason: "unknown_node" };
    this.chargesUsed += 1;
    if (this.model.isWon()) {
      this.status = "won";
      this.wonBoardIds.add(this.board.id);
    } else if (this.chargesUsed >= this.board.charges) {
      this.status = "exhausted";
    }
    return { accepted: true };
  }

  /** 三板都曾被打通才算整关完成；不信任 nextBoard() 调用顺序。 */
  isLevelComplete(): boolean {
    return this.boards.every((board) => this.wonBoardIds.has(board.id));
  }

  /** 重来本板(电量、状态、星网清零) */
  resetBoard(): void {
    this.model.reset();
    this.chargesUsed = 0;
    this.status = "playing";
    this.wonBoardIds.delete(this.board.id);
  }

  /** 进入下一板; 已是最后一板返回 false */
  nextBoard(): boolean {
    if (this.boardIndex >= this.boards.length - 1) return false;
    this.boardIndex += 1;
    this.model = this.buildModel(this.boardIndex);
    this.chargesUsed = 0;
    this.status = "playing";
    return true;
  }

  get view(): StarWebView {
    const board = this.board;
    const nodes: StarNodeView[] = board.layout.nodes.map((node) => {
      const life = this.model.lifeOf(node.id);
      const lit = life > 0;
      const frozen = lit && this.model.litNeighborCount(node.id) >= this.rules.freezeThreshold;
      const status: StarNodeStatus = !lit ? "dark" : frozen ? "frozen" : "decaying";
      return { id: node.id, x: node.x, y: node.y, life, lit, status };
    });
    return {
      boardId: board.id,
      boardIndex: this.boardIndex,
      boardCount: this.boards.length,
      nodes,
      edges: board.layout.edges,
      chargesTotal: board.charges,
      chargesLeft: Math.max(0, board.charges - this.chargesUsed),
      status: this.status
    };
  }

  private get board(): StarBoard {
    return this.boards[this.boardIndex];
  }

  private buildModel(index: number): StarNetworkModel {
    return new StarNetworkModel(boardGraph(this.boards[index]), this.rules);
  }
}

File: StarGuardian/Assets/Scripts/M02/M02StarWebSession.cs
// M02《点亮你温暖我》会话层 —— 纯逻辑(无 UnityEngine), 从 assets/scripts/cocos/M02StarWebSession.ts 逐字迁移, 规则不变.
// 管三件 domain 之上的东西: 电量(charges)、三板推进、每颗星的呈现态 + 胜负。
// 引擎胶水层(视图组件)只读 View、把 tap 转给 TapNode(), 不自己算规则。
#nullable enable

using System;
using System.Collections.Generic;
using System.Linq;
using StarGuardian.Core;

namespace StarGuardian.M02
{
    /// <summary>单颗星的呈现态: 暗 / 衰减中(亮但支撑不足) / 冻结(亮且亮邻居达标)
    /// —— TS `type StarNodeStatus` 字符串联合(沿用 M01 惯例落成 const string, 不建 enum, 测试断言依赖字面值)</summary>
    public static class StarNodeStatus
    {
        public const string Dark = "dark";
        public const string Decaying = "decaying";
        public const string Frozen = "frozen";
    }

    /// <summary>一板的状态: 进行中 / 已全锁胜利 / 电量耗尽未胜 —— TS `type BoardStatus` 字符串联合</summary>
    public static class BoardStatus
    {
        public const string Playing = "playing";
        public const string Won = "won";
        public const string Exhausted = "exhausted";
    }

    /// <summary>TapResult.Reason 的取值 —— TS reason?: "not_playing" | "unknown_node"</summary>
    public static class TapRejectReason
    {
        public const string NotPlaying = "not_playing";
        public const string UnknownNode = "unknown_node";
    }

    public sealed class StarNodeView
    {
        public string Id { get; init; } = "";
        public double X { get; init; }
        public double Y { get; init; }
        /// <summary>TS number; 恒为整数拍(来源 StarNetworkModel.LifeOf, int) → int</summary>
        public int Life { get; init; }
        public bool Lit { get; init; }
        public string Status { get; init; } = StarNodeStatus.Dark;
    }

    public sealed class StarWebView
    {
        public string BoardId { get; init; } = "";
        public int BoardIndex { get; init; }
        public int BoardCount { get; init; }
        public IReadOnlyList<StarNodeView> Nodes { get; init; } = Array.Empty<StarNodeView>();
        /// <summary>TS ReadonlyArray&lt;readonly [string, string]&gt; —— 与 TS 同为透传 board.Layout.Edges 的共享引用(协变只读视图)</summary>
        public IReadOnlyList<IReadOnlyList<string>> Edges { get; init; } = Array.Empty<IReadOnlyList<string>>();
        public int ChargesTotal { get; init; }
        public int ChargesLeft { get; init; }
        public string Status { get; init; } = BoardStatus.Playing;
    }

    public sealed class TapResult
    {
        public bool Accepted { get; init; }
        /// <summary>TS reason?: 缺省(undefined)落成 null; 取值见 TapRejectReason</summary>
        public string? Reason { get; init; }
    }

    public sealed class StarWebSession
    {
        private readonly List<StarBoard> boards;
        private readonly StarNetworkRules rules;
        private int boardIndex = 0;
        private StarNetworkModel model;
        private int chargesUsed = 0;
        private string status = BoardStatus.Playing;
        private readonly HashSet<string> wonBoardIds = new();

        public StarWebSession(StarWebConfig config)
        {
            if (config.Boards.Count == 0)
            {
                // TS 行 49: throw new Error(...) —— 本仓惯例落成 InvalidOperationException, 文案逐字
                throw new InvalidOperationException("StarWebSession requires at least one board");
            }
            boards = config.Boards; // TS 行 51: 共享引用, 不拷贝
            rules = config.Mechanic; // TS 行 52: mechanic 结构化直传 → C# 隐式转换 StarWebMechanic→StarNetworkRules
            model = BuildModel(0);
        }

        /// <summary>点一颗星 = 花一点电量走一拍。未知星不消耗电量、不推进；非进行中拒绝。</summary>
        public TapResult TapNode(string id)
        {
            if (status != BoardStatus.Playing)
            {
                return new TapResult { Accepted = false, Reason = TapRejectReason.NotPlaying };
            }
            if (!model.Step(id))
            {
                return new TapResult { Accepted = false, Reason = TapRejectReason.UnknownNode };
            }
            chargesUsed += 1;
            if (model.IsWon())
            {
                status = BoardStatus.Won;
                wonBoardIds.Add(Board.Id);
            }
            else if (chargesUsed >= Board.Charges)
            {
                status = BoardStatus.Exhausted;
            }
            return new TapResult { Accepted = true };
        }

        /// <summary>三板都曾被打通才算整关完成；不信任 NextBoard() 调用顺序。</summary>
        public bool IsLevelComplete()
        {
            return boards.All(board => wonBoardIds.Contains(board.Id)); // TS 行 72: .every
        }

        /// <summary>重来本板(电量、状态、星网清零)</summary>
        public void ResetBoard()
        {
            model.Reset();
            chargesUsed = 0;
            status = BoardStatus.Playing;
            wonBoardIds.Remove(Board.Id); // TS 行 80: Set.delete
        }

        /// <summary>进入下一板; 已是最后一板返回 false</summary>
        public bool NextBoard()
        {
            if (boardIndex >= boards.Count - 1) return false;
            boardIndex += 1;
            model = BuildModel(boardIndex);
            chargesUsed = 0;
            status = BoardStatus.Playing;
            return true;
        }

        public StarWebView View
        {
            get
            {
                var board = Board;
                // TS 行 95-101: board.layout.nodes.map(...)
                var nodes = board.Layout.Nodes.Select(node =>
                {
                    var life = model.LifeOf(node.Id);
                    var lit = life > 0;
                    var frozen = lit && model.LitNeighborCount(node.Id) >= rules.FreezeThreshold;
                    var nodeStatus = !lit ? StarNodeStatus.Dark : frozen ? StarNodeStatus.Frozen : StarNodeStatus.Decaying;
                    return new StarNodeView { Id = node.Id, X = node.X, Y = node.Y, Life = life, Lit = lit, Status = nodeStatus };
                }).ToList();
                return new StarWebView
                {
                    BoardId = board.Id,
                    BoardIndex = boardIndex,
                    BoardCount = boards.Count,
                    Nodes = nodes,
                    Edges = board.Layout.Edges,
                    ChargesTotal = board.Charges,
                    ChargesLeft = Math.Max(0, board.Charges - chargesUsed), // TS 行 109: Math.max(0, ...)
                    Status = status
                };
            }
        }

        private StarBoard Board => boards[boardIndex];

        private StarNetworkModel BuildModel(int index)
        {
            // TS 行 119: boardGraph(this.boards[index]) → C# StarWebConfigValidator.BoardGraphOf
            return new StarNetworkModel(StarWebConfigValidator.BoardGraphOf(boards[index]), rules);
        }
    }
}
